### PR TITLE
Import pretained python AC model to C++

### PR DESCRIPTION
**Pull request by jakesabathia2**
#2378

### CODE REVIEW

**Comment by jakesabathia2** [comment]
So I just find out the evaluation need random seed anyway,
so I just generate a random seed when importing the old Python model.

File: src/ml/neural_net/weight_init.cpp [hunk 1]
```diff
@@ -61,7 +61,29 @@ xavier_weight_initializer::xavier_weight_initializer(
   result.output_gate_bias_fn = zero_weight_initializer();
 
   return result;
-}
+  }
```
Comment (nickjong) [inline src/ml/neural_net/weight_init.cpp]: This curly brace is now mis-aligned. Please consider running git-clang-format (just on your changes)

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 2]
```diff
@@ -33,6 +33,7 @@ using neural_net::lstm_weight_initializers;
 using neural_net::shared_float_array;
 using neural_net::xavier_weight_initializer;
 using neural_net::zero_weight_initializer;
+using neural_net::weight_initializer;
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: Nit: let's keep these in alphabetical order. (clang-format may or may not do this automatically for you)

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 3]
```diff
@@ -524,6 +525,11 @@ void activity_classifier::import_from_custom_model(
   flex_dict pred_model = variant_get_value<flex_dict>(it->second);
   model_data.erase(it);
 
+  //generate a random seed for the model
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: Why is a random seed required for evaluation?

File: src/toolkits/activity_classification/activity_classifier.hpp [hunk 4]
```diff
@@ -201,7 +201,7 @@ class EXPORT activity_classifier: public ml_model_base {
       const;
 
   // Returns the initial neural network to train
-  virtual std::unique_ptr<neural_net::model_spec> init_model() const;
+  virtual std::unique_ptr<neural_net::model_spec> init_model(bool use_zero_init = false) const;
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.hpp]: We shouldn't use default arguments for virtual functions, due to nonintuitive semantics. See also https://google.github.io/styleguide/cppguide.html#Default_Arguments

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 5]
```diff
@@ -652,15 +659,23 @@ std::unique_ptr<model_spec> activity_classifier::init_model() const
   size_t prediction_window = read_state<flex_int>("prediction_window");
   const flex_list &features_list = read_state<flex_list>("features");
 
-  // Initialize a random number generator for weight initialization.
-  std::seed_seq seed_seq = { read_state<int>("random_seed") };
-  std::mt19937 random_engine(seed_seq);
-
+  //Only to create a random engine for weight initialization if use_zero_init = false
+  std::mt19937 random_engine;
+  if (use_zero_init == false){
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: Our code tends to the style `if (!use_zero_init) {`

Might be more clear to use the inverse: `use_random_init`

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 6]
```diff
@@ -652,15 +659,23 @@ std::unique_ptr<model_spec> activity_classifier::init_model() const
...
+    std::seed_seq seed_seq { read_state<int>("random_seed") };
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: Let's resolve whether we really need random seed or not for evaluation. If we do, there's no need to make this conditional: we can always initialize the random_engine from the random seed

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 7]
```diff
@@ -652,15 +659,23 @@ std::unique_ptr<model_spec> activity_classifier::init_model() const
...
       "features",
       std::vector<std::string>(features_list.begin(), features_list.end()));
   result->add_reshape("reshape", "features",
                       {{1, num_features, 1, prediction_window}});
+  weight_initializer initializer;
+  lstm_weight_initializers lstm_initializer;
+  initializer = (use_zero_init) ? (weight_initializer) zero_weight_initializer() :
+  (weight_initializer) xavier_weight_initializer(num_features * prediction_window,
+                                NUM_CONV_FILTERS * prediction_window,
+                                &random_engine);
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: I think you can just initialize `initializer` to a `zero_weight_initializer`, then only overwrite it when use random initialization.

Then we can avoid the unnecessary use of the ternary ?: operator

File: src/toolkits/activity_classification/activity_classifier.cpp (hunk 3 above)
Comment (jakesabathia2) [reply src/toolkits/activity_classification/activity_classifier.cpp]: it somehow need to sample the time series data with a random seed.

File: src/toolkits/activity_classification/activity_classifier.cpp (hunk 6 above)
Comment (jakesabathia2) [reply src/toolkits/activity_classification/activity_classifier.cpp]: So the evaluation always need a random seed.
However, I guess in here we want to use the previous produced random seed 
from the saved model to reproduce the result.
Hmmm so maybe I should change the name of the bool.

File: src/ml/neural_net/weight_init.cpp [hunk 8]
```diff
@@ -40,49 +40,49 @@ xavier_weight_initializer::xavier_weight_initializer(
...
-  result.forget_gate_bias_fn = zero_weight_initializer();
-  result.block_input_bias_fn = zero_weight_initializer();
-  result.output_gate_bias_fn = zero_weight_initializer();
+    result.input_gate_bias_fn = zero_weight_initializer();
+    result.forget_gate_bias_fn = zero_weight_initializer();
+    result.block_input_bias_fn = zero_weight_initializer();
+    result.output_gate_bias_fn = zero_weight_initializer();
 
-  return result;
+    return result;
```
Comment (nickjong) [inline src/ml/neural_net/weight_init.cpp]: These lines should not be indented further. Is this from clang-format? It looks like it's being thrown off by the incorrect indentation in line 31.

Don't just blindly accept what clang-format outputs!

File: src/toolkits/activity_classification/activity_classifier.cpp [hunk 9]
```diff
@@ -639,7 +635,12 @@ std::unique_ptr<data_iterator> activity_classifier::create_iterator(
   data_params.prediction_window = read_state<flex_int>("prediction_window");
   data_params.predictions_in_chunk = NUM_PREDICTIONS_PER_CHUNK;
   data_params.use_data_augmentation = use_data_augmentation;
-  data_params.random_seed = read_state<int>("random_seed");
+  if (!use_data_augmentation){
+    data_params.random_seed = 0;
+  }
+  else{
+    data_params.random_seed = read_state<int>("random_seed");
+  }
```
Comment (nickjong) [inline src/toolkits/activity_classification/activity_classifier.cpp]: Did you run clang-format here? In most of our code, the formatting should be
```
if (condition) {
  ...
} else {
  ...
}
```

Comment (jakesabathia2) [reply src/toolkits/activity_classification/activity_classifier.cpp]: hmmm it seems like the clang-format didn't run when I commit it.
Will take a look

**Pull request closed without merging**